plugins/scaffolder-backend/src/scaffolder/actions/builtin/publish/bitbucket.ts

```diff
@@ -0,0 +1,251 @@
+};
+
+const createBitbucketServerRepository = async (opts: {
+  host: string;
+  owner: string;
+  repo: string;
+  description: string;
+  repoVisibility: 'private' | 'public';
+  authorization: string;
+}) => {
+  const {
+    host,
+    owner,
+    repo,
+    description,
+    authorization,
+    repoVisibility,
+  } = opts;
+
+  let response: Response;
+  const options: RequestInit = {
+    method: 'POST',
+    body: JSON.stringify({
+      name: repo,
+      description: description,
+      is_private: repoVisibility === 'private',
+    }),
+    headers: {
+      Authorization: authorization,
+      'Content-Type': 'application/json',
+    },
+  };
+
+  try {
+    response = await fetch(
+      `https://${host}/rest/api/1.0/projects/${owner}/repos`,
+      options,
+    );
+  } catch (e) {
+    throw new Error(`Unable to create repository, ${e}`);
+  }
+
+  if (response.status !== 201) {
+    throw new Error(
+      `Unable to create repository, ${response.status} ${
+        response.statusText
+      }, ${await response.text()}`,
+    );
+  }
+
+  const r = await response.json();
+  let remoteUrl = '';
+  for (const link of r.links.clone) {
+    if (link.name === 'http') {
+      remoteUrl = link.href;
+    }
+  }
+
+  const repoContentsUrl = `${r.links.self[0].href}`;
+  return { remoteUrl, repoContentsUrl };
+};
+
+const getAuthorizationHeader = (config: BitbucketIntegrationConfig) => {
+  if (config.username && config.appPassword) {
+    const buffer = Buffer.from(
+      `${config.username}:${config.appPassword}`,
+      'utf8',
+    );
+
+    return `Basic ${buffer.toString('base64')}`;
+  }
+
+  if (config.token) {
+    return `Bearer ${config.token}`;
+  }
+
+  throw new Error(
+    `Authorization has not been provided for Bitbucket. Please add either username + appPassword or token to the Integrations config`,
+  );
+};
+
+export function createPublishBitbucketAction(options: {
+  integrations: ScmIntegrations;
+}) {
+  const { integrations } = options;
+
+  return createTemplateAction<{
+    repoUrl: string;
+    description: string;
+    repoVisibility: 'private' | 'public';
+  }>({
+    id: 'publish:bitbucket',
+    schema: {
+      input: {
+        type: 'object',
+        required: ['repoUrl'],
+        properties: {
+          repoUrl: {
+            title: 'Repository Location',
+            type: 'string',
+          },
+          description: {
+            title: 'Repository Description',
+            type: 'string',
+          },
+          repoVisibility: {
+            title: 'Repository Visiblity',
+            type: 'string',
+            enum: ['private', 'public'],
+          },
+        },
+      },
+      output: {
+        type: 'object',
+        properties: {
+          remoteUrl: {
+            title: 'A URL to the repository with the provider',
+            type: 'string',
+          },
+          repoContentsUrl: {
+            title: 'A URL to the root of the repository',
```

Comment:
These two are confusingly worded. If you take the descriptions out of their context,

"A URL to the repository with the provider" - not clear to me what this means

"A URL to the root of the repository" - how is this even different from the one above

Same for the other schemas